hoomd/mpcd/geometry.py

```diff
@@ -56,6 +56,128 @@ def __init__(self, no_slip):
         self._param_dict.update(param_dict)
 
 
+class CosineChannel(Geometry):
+    r"""Serpentine (sinusoidal) channel.
+
+    Args:
+        amplitude (float): Amplitude of cosine.
+        repeat_length (float): Repeat length (period) of cosine.
+        separation (float): Distance between channel walls.
+        no_slip (bool): If True, surfaces have no-slip boundary condition.
+            Otherwise, they have the slip boundary condition.
+
+    `CosineChannel` models a fluid confined in :math:`y` between two
+    walls described by a sinusoidal profile with equations
+
+    .. math::
+
+        y(x) = A \cos\left(\frac{2 \pi x}{L}\right) \pm H
+
+    where :math:`A` is the `amplitude`, :math:`L` is the `repeat_length`,
+    and :math:`2H` is the `separation`.
+
+    .. rubric:: Example:
+
+    .. code-block:: python
+
+        channel = hoomd.mpcd.geometry.CosineChannel(
+            amplitude=2.0,
+            separation=4.0,
+            repeat_length=10.0)
+        stream = hoomd.mpcd.stream.BounceBack(period=1, geometry=channel)
+        simulation.operations.integrator.streaming_method = stream
+
+    Attributes:
+        amplitude (float): Amplitude of cosine (*read only*).
+
+        repeat_length (float): Repeat length (period) of cosine. (*read only*).
+
+        separation (float): Distance between walls (*read only*).
+
+
+    """
+
+    def __init__(self, amplitude, repeat_length, separation, no_slip=True):
+        super().__init__(no_slip)
+
+        param_dict = ParameterDict(
+            amplitude=float(amplitude),
+            repeat_length=float(repeat_length),
+            separation=float(separation),
+        )
+        self._param_dict.update(param_dict)
+
+    def _attach_hook(self):
+        self._cpp_obj = _mpcd.CosineChannel(self.amplitude, self.repeat_length,
+                                            self.separation, self.no_slip)
+        super()._attach_hook()
+
+
+class CosineExpansionContraction(Geometry):
+    r"""Channel with sinusoidal expansion and contraction.
+
+    Args:
+        expansion_separation (float): Maximum distance between channel walls.
+        contraction_separation (float): Minimum distance between channel walls.
+        repeat_length (float): Repeat length (period) of cosine.
+        no_slip (bool): If True, surfaces have no-slip boundary condition.
+            Otherwise, they have the slip boundary condition.
+
+    `CosineExpansionContraction` models a fluid confined in :math:`y` between
+    two walls described by a sinusoidal profile with equations
+
+    .. math::
+
+        y(x) = \pm\left( \frac{H_{\rm e}-H_{\rm c}}{2}
```

Comment:
I think there is a 1/2 missing in this equation.

Reply:
Did you account for `H_e` and `H_c` being defined as half `expansion_separation` and `contraction_separation` below? This is what I get if I use this equation for the example below, with `He = 0.5 * 6.0` and `Hc = 0.5 * 3.0`

![image](https://github.com/user-attachments/assets/da64c258-f73e-4350-9468-75233cb78a97)